\documentclass[12pt]{article}
\usepackage[a4paper,top=20mm,bottom=20mm,left=20mm,right=20mm]{geometry}
\usepackage{url}
\usepackage{alltt}
\usepackage{xspace}
\usepackage{times}
\usepackage{bbm}
\usepackage{verbatim}
\usepackage{ifthen}
\usepackage{optionman}

\newcommand{\Substring}[3]{#1[#2..#3]}

\newcommand{\Program}[0]{\texttt{tagerator}\xspace}

\title{\Program: a program for mapping short sequence tags\\
       a manual}

\author{\begin{tabular}{c}
         \textit{Stefan Kurtz}\\
         Center for Bioinformatics,\\
         University of Hamburg
        \end{tabular}}

\begin{document}
\maketitle

\section{Preliminary definitions}
By \(S\) let us denote the concatenation of all subject sequences.
By \(edist(u,v)\) we denote the unit edit distance  of two strings
\(u\) and \(v\).
Consider a sequence \(p\) of length \(m\). An approximate match of \(p\) with 
up to \(k\) differences is a substring \(v\) of \(S\) such that 
\(edist(p,v)\leq k\). An approximate prefix match of \(p\) with \(k\) 
differences and at most \(t\) occurrences is a substring \(v\) of \(S\) such 
that \(v\) occurs at most \(t\) times in \(S\) and 
\(edist(v,\Substring{p}{1}{i})\leq k\) for some \(i\in[1,m]\).

\section{The program \Program}

The program \Program is called as follows:
\par
\noindent\texttt{gt} \Program [\textit{options}] \Showoption{q} \Showoptionarg{files} [\textit{options}] 
\par
\Showoptionarg{files} is a white space separated list of at least one
filename. Any sequence occurring in any file specified in
\Showoptionarg{files} is called \textit{short sequence tag} or \textit{tag}
for short. In addition to the mandatory option \Showoption{q}, the program
must be called with either option \Showoption{pck} or \Showoption{esa},
which specify to use a packed index and an enhanced suffix array,
respectively. Both indices are constructed from a given set of subject 
sequences. 

\Program maps each short sequence tag, say \(p\) of length \(m\)
against the given index. The length of the tag is limited by the
size of a pointer. If \texttt{gt} is a 32-bit binary, then \(m\) must be
smaller or equal to 32. If \texttt{gt} is a 64-bit binary, \(m\) must be 
smaller or equal to 64. The program runs in three basic modes:
\begin{itemize}
\item
In the \textit{ms}-mode, it computes for all \(i\in[1,m]\) the length \(\ell\) 
of the longest prefix of \(\Substring{w}{i}{m}\) matching any substring of 
\(S\).
In addition, it reports a start position, say \(j\), of such a prefix in \(S\).
As these values make up the well known matching statistics, we denote 
it by \textit{ms}. The length value \(\ell\) is the matching statistics length
and the position value \(j\) is a matching statistics position.
While the matching statistics length is uniquely determined, the 
matching statistics position is not.
\item
In the \textit{cdiff}-mode, it computes all start positions of approximate 
matches with up to \(k\) differences in \(S\). For each start position of 
an approximate match, say \(j\), it reports the minimum integer \(\ell\) such 
that \(edist(p,\Substring{S}{j}{j+\ell-1})\leq k\). Since this mode matches
the complete sequence \(p\), we call this mode \textit{cdiff},
for complete difference.
\item
In the \textit{pdiff}-mode, it computes all start positions of approximate 
prefix matches with up to \(k\) differences and at most \(t\) occurrences in 
\(S\). For each start position of an approximate prefix match, say \(j\), it 
reports the minimum integers \(i\) and \(\ell\) such that 
\(edist(\Substring{p}{1}{i},\Substring{S}{j}{j+\ell-1})\leq k\). Since this 
mode matches a prefix of \(p\) with some differences, we call this mode 
\textit{pdiff}.
\end{itemize}

The following options are available in \Program:

\begin{Justshowoptions}
\Option{q}{$\Showoptionarg{files}$}{
Specify a white space separated list of query files (in multiple \Fasta format)
containing the tags. At least one query file must be given. The files may be in 
gzipped format, in which case they have to end with the suffix \texttt{.gz}.
}

\Option{esa}{$\Showoptionarg{indexname}$}{
Use the given enhanced suffix array index to map the short sequence tags.
}

\Option{pck}{$\Showoptionarg{indexname}$}{
Use the packed index (an efficient representation of the FMindex)
to map the short sequence tags.
}

\Option{e}{$\Showoptionarg{k}$}{
Specify the number of differences allowed. \(k\) must be a non-negative number
\(k=0\) means that no differences are allowed (exact matching) and \(k>0\)
means a positive number of differences. If this option is not used, then
the program runs in \textit{ms}-mode, i.e.\ it computes the matching statistics
for each short sequence tag.
}

\Option{d}{}{
Compute direct matches. This is the default case}

\Option{p}{}{
Compute palindromic matches, i.e.\ reverse complemented matches.
This option does not work currently.}

\Option{maxocc}{$\Showoptionarg{t}$}{
Specify the maximum number of occurrences of approximate prefix matches.
}

\Option{nowildcards}{}{
Do not output matches containing wildcard characters (e.g. N). This
option is not relevant for the \textit{ms}-mode and for the other modes
with \(k=0\).
}

\Helpoption

\end{Justshowoptions}
The following conditions must be satisfied:
\begin{enumerate}
\item
Option \Showoption{q} is mandatory.
\item
Either option \Showoption{pck} or \Showoption{esa} must be used. Both cannot
be combined.
\item
If option \Showoption{e} is not used, then option \Showoption{maxocc}
is required.
\end{enumerate}

\section{Examples}
Suppose that in some directory, say \texttt{homo-sapiens}, we have 24 gzipped
\Fasta files containing all 24 human chromomsomes. These may have been 
downloaded from
\url{ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens_47_36i/dna}.

In the first step, we construct the packed index for the entire human genome:

\begin{Output}
gt packedindex mkindex -dna -dir rev -parts 12 -bsize 10 -sprank -locfreq 32
                       -indexname pck-human -db homo-sapiens/*.gz
\end{Output}

The program runs for little more than two hours and delivers 
an index \texttt{human-all} consisting of three files:

\begin{Output}
ls -lh human-all.*
-rw-r----- 1 kurtz gistaff   37 2008-07-13 17:00 pck-human.al1
-rw-r----- 1 kurtz gistaff 2.6G 2008-07-13 19:22 pck-human.bdx
-rw-r----- 1 kurtz gistaff 3.3K 2008-07-13 19:22 pck-human.prj
\end{Output}

Suppose that the compressed file \texttt{Q1.gz} contains short sequence tags
in multiple \Fasta format. In a first call we run \Program in \textit{ms}-mode:

\begin{Output}
gt tagerator -q Q1.gz -pck pck-human
# computing matching statistics
# indexname(pck)=pck-human
# queryfile=Q1.gz
# 15 aagcttgctgctgca
10 88693
9 88694
8 88695
12 470064
11 470065
10 729958
9 772093
8 403975
7 544167
6 680243
5 714382
4 313675
3 185058
2 739095
1 274
...
\end{Output}
The first line shortly reports the kind of computation performed. The second
and third line give the name of the index containing the subject sequences.
It is reported wether it is a packed index or an enhanced suffix array. Then for
each tag its length, say \(m\), and the tag \(p\) itself is shown, followed by 
a block of \(m\) lines each containing two integers. For all \(i\in[1,m]\), 
the first number in the \(i\)th line is the matching statistics length,
and the second number in the \(i\)th line is a matching statistics position.
So, for example, the sequence \texttt{gcttgctg} of length 8 starting at tag
position 3 is the longest prefix of \texttt{gcttgctgctgca} that occurs as 
a substring in the index. It, for example, occurs at position 88695.

In a second call, we run \Program in \textit{cdiff}-mode:

\begin{Output}
gt tagerator -q Q1.gz -pck pck-human -e 2 -nowildcards
# computing complete matches with up to 2 differences
# indexname(pck)=pck-human
# queryfile=Q1.gz
# 15 aagcttgctgctgca
14 363967
13 576657
13 501200
13 158315
13 180391
15 464011
14 88697
13 576546
15 191866
15 135753
15 668839
13 576656
13 501199
14 471053
14 78690
14 104932
14 463357
14 317582
15 576655
15 501198
15 317582
...
\end{Output}
The first line shortly reports the kind of computation performed. The second
and third line are as in the previous example. Then for
each tag its length and the tag itself is shown, followed by 
a block of lines each containing a pair of integers. Each such pair
reports the length of an approximate match and its start position in the
subject sequence \(S\). For each
start position only the shortest length is reported. If there are no
approximate matches, then no such line appears.

Our third example concerns the \textit{pdiff}-mode:

\begin{Output}
# computing prefix matches <= 1 differences and at most 5 occurrences in the subject sequences
# indexname(pck)=pck-human
# queryfile=Q1.gz
# 15 aagcttgctgctgca
8 507736 8
8 213544 8
8 289017 8
9 274066 8
9 427225 8
9 632436 9
9 590998 9
9 411370 9
9 281698 9
9 367734 9
9 331595 9
9 342090 9
8 616567 7
8 507735 7
8 591679 7
9 88694 10
...
\end{Output}

The output is similar as in the previous example, except that each match
is reported by three integers. The first number reports the length of the
approximate match, the second reports the start position of the approximate
match in the subject sequence and the third reports the length of the prefix 
of the tag involved in the approximate match. Note that in some cases,
the two length values are different.
\end{document}
