\documentclass[12pt]{article}
\usepackage[a4paper,top=20mm,bottom=20mm,left=20mm,right=20mm]{geometry}
\usepackage{url}
\usepackage{alltt}
\usepackage{xspace}
\usepackage{times}
\usepackage{listings}
\usepackage{bbm}
\usepackage{verbatim}
\usepackage{ifthen}
\usepackage{comment}
\usepackage{optionman}

\newcommand{\Substring}[3]{#1[#2..#3]}

\newcommand{\Program}[0]{\texttt{tagerator}\xspace}

\title{\Program: a program for mapping short sequence tags\\
       a manual}

\author{\begin{tabular}{c}
         \textit{Stefan Kurtz}\\
         Center for Bioinformatics,\\
         University of Hamburg
        \end{tabular}}

\begin{document}
\maketitle

\section{Preliminary definitions}
By \(S\) let us denote the concatenation of all subject sequences.
By \(edist(u,v)\) we denote the unit edit distance  of \(u\) and \(v\).
Consider a sequence \(p\) of length \(m\). An approximate match of \(p\) with 
up to \(k\) differences is a substring \(v\) of \(S\) such that 
\(edist(p,v)\leq k\). An approximate prefix match of \(p\) with \(k\) 
differences and at most \(t\) occurrences is a substring \(v\) of \(S\) such 
that \(v\) occurs at most \(t\) times in \(S\) and 
\(edist(v,\Substring{p}{1}{i})\leq k\) for some \(i\in[1,m]\).

\section{The program \Program}

The program \Program is called as follows:
\par
\noindent\Program [\textit{options}] \Showoption{query} \Showoptionarg{files} [\textit{options}] 
\par
\Showoptionarg{files} is a white space separated list of at least one
filename. Any sequence occurring in any file specified in
\Showoptionarg{files} is called \textit{short sequence tag} or \textit{tag}
for short. In addition to the mandatory option \Showoption{query}, the program
must be called with either option \Showoption{pck} or \Showoption{esa},
which specify to use a packed index and an enhanced suffix array,
respectively. Both indices are constructed from a given set of subject 
sequences. 

\Program maps each short sequence tag, say \(p\) of length \(m\)
against the given index. The program runs in three basic modes:
\begin{description}
\item[ms]
For all \(i\in[1,m]\) it can compute the length of the longest common prefix 
of \(\Substring{w}{i}{m}\) matching any substring of \(S\).
In addition it reports a start position of such a prefix in \(S\).
As this mode basically computes the well known matching statistics, we denote 
it by \textit{ms}.
\item[cdiff]
It computes all start positions of approximate matches with up to \(k\)
differences in \(S\). For each start position of an approximate match,
say \(j\), it the minimum integer \(\ell\) such that
\(edist(\Substring{S}{j}{j+\ell-1},p)\leq k\). Since this mode matches
the complete sequence \(p\), we call this mode \textit{cdiff},
for complete difference.
\item[pdiff]
It computes all start positions of approximate prefix matches with up to \(k\)
differences and at most \(t\) occurrences in \(S\). For each start position 
of an approximate prefix match, say \(j\), it reports minimum integer \(i\) and
\(\ell\) such that 
\(edist(\Substring{S}{j}{j+\ell-1},\Substring{p}{1}{i})\leq k\). Since this 
mode matches a prefix of \(p\) with some differences, we call this mode 
\textit{pdiff}.
\end{description}

The following options are available in \Program:

\begin{Justshowoptions}
\Option{query}{$\Showoptionarg{files}$}{
Specify a white space separated list of query files (in multiple \Fasta format)
containing the tags. At least one query file must be given. The files may be in 
gzipped format, in which case they have to end with the suffix \texttt{.gz}.
}

\Option{esa}{$\Showoptionarg{indexname}$}{
Use the given enhanced suffix array index to map the short sequence tags.
}

\Option{pck}{$\Showoptionarg{indexname}$}{
Use the packed index (an efficient representation of the FMindex)
to map the short sequence tags.
}

\Option{e}{$\Showoptionarg{k}$}{
Specify the number of differences allowed. \(k\) must be a non-negative number
\(k=0\) means that no differences are allowed (exact matching) and \(k>0\)
means a positive number of differences. If this option is not used, then
the program runs in \textit{ms}-mode, i.e.\ it computes the matching statistics
for each short sequence tag.
}

\Option{d}{}{
Compute direct matches. This is the default case}

\Option{p}{}{
Compute palindromic matches, i.e.\ reverse complemented matches.
This option does not work currently.}

\Option{maxocc}{t}{
Specify the maximum number of occurrences of approximate prefix matches.
}

\Option{nospecials}{}{
Do not output matches cont containing wildcard characters (e.g. N). This
option is not relevant for the \textit{ms}-mode.
}

\Helpoption

\end{Justshowoptions}
The following conditions must be satisfied:
\begin{enumerate}
\item
Option \Showoption{query} is mandatory.
\item
Either option \Showoption{pck} or \Showoption{esa} must be used. Both cannot
be combined.
\item
Option \Showoption{maxocc} can only be used in combination with 
option \Showoption{e}.
\end{enumerate}

\section{Examples}
Suppose that in some directory, say \texttt{homo-sapiens}, we have 24 gzipped
\Fasta files containing all 24 human chromomsomes. These may have been 
downloaded from
\url{ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens_47_36i/dna}.

In the first step, we construct the packed index for the entire human genome:

\begin{Output}
gt packedindex mkindex -dna -dir rev -parts 12 -bsize 10 -sprank -locfreq 32
                       -indexname pck-human -db homo-sapiens/*.gz
\end{Output}

The program runs for little more than two hours and delivers 
an index \texttt{human-all} consisting of three files:

\begin{Output}
ls -lh human-all.*
-rw-r----- 1 kurtz gistaff   37 2008-07-13 17:00 pck-human.al1
-rw-r----- 1 kurtz gistaff 2.6G 2008-07-13 19:22 pck-human.bdx
-rw-r----- 1 kurtz gistaff 3.3K 2008-07-13 19:22 pck-human.prj
\end{Output}

Suppose that the compressed file \texttt{Q1.gz} contains short sequence tags
in multiple \Fasta format. In a first call we run \Program in \textit{ms}-mode:

\begin{Output}
gt tagerator -query Q1.gz -pck pck-human
# queryfile=Q1.gz
# computing matching statistics
# indexname(pck)=pck-human
# patternlength=15
# tag=aagcttgctgctgca
0->10
1->9
2->8
3->12
4->11
5->10
6->9
7->8
8->7
9->6
10->5
11->4
12->3
13->2
14->1
...
\end{Output}
The first
\end{document}

For all short sequence tag in the multiple \Fasta file \texttt{queryfile.fna},
a line is shown, reporting the number of the unit and the original \Fasta
header. Also, all for positions \(i\) in \(s\) satisfying 
\(20\leq \Mup{i}\leq 30\), \(i\) and \(\Mup{i}\) is reported.

The first column is the relative position in the unit sequence (counting
from 0). The second column shows the length value.

To additionally report the sequence content of the
minimum unique prefixes we add the keyword \Showoptionkey{sequence} to option
\Showoption{output}:

\begin{Output}
gt uniquesub -output querypos sequence -min 20 -max 30 
             -query queryfile.fna -pck human-all
unit 0 (Mus musculus, chr 1, complete sequence)
1007 20 ctgacagtttttttttttta
1010 22 acagttttttttttttacttta
1011 22 cagttttttttttttactttat
1012 21 agttttttttttttactttat
1013 21 gttttttttttttactttata
...
\end{Output}

\begin{Output}
gt matstat -output subjectpos querypos sequence -min 20 -max 30 
           -query queryfile.fna -pck human-all
unit 0 (Mus musculus, chr 1, complete sequence)
22 20 390765125 actgtatctcaaaatataaa
253 21 258488266 gggaataaacatgtcattgag
254 20 258488267 ggaataaacatgtcattgag
275 20 900483549 taattctatttttctttctt
480 20 1008274536 gcttgaagatcatgatccag
..
\end{Output}
Here, the first column shows the relative positions in unit 0 for which the
length of the matching statistics is between 20 and 30. The second column is
the corresponding length value. The third column shows position of the
matching sequence in the index, and the fourth shows the sequence content.
\end{document}
