<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta http-equiv="content-type" content="text/html; charset=iso-8859-1">
<title>GenomeTools</title>
<link rel="stylesheet" type="text/css" href="style.css">
</head>
<body>
  <div id="menu">
    <ul>
      <li><a id="current" href="index.html">Overview</a></li>
      <li><a href="pub/">Download</a></li>
      <li><a href="cgi-bin/gitweb.cgi?p=genometools.git;a=summary">Browse source</a></li>
      <li><a href="mailman/listinfo/gt-users">Mailing list</a></li>
      <li><a href="http://genometools.lighthouseapp.com/">Issue tracker</a></li>
      <li><a href="design.html">Design</a></li>
      <li><a href="libgenometools.html">C API</a></li>
      <li><a href="docs.html"><tt>gtscript</tt> docs</a></li>
      <li><a href="annotationsketch.html"><tt>AnnotationSketch</tt></a></li>
      <li><a href="license.html">License</a></li>
    </ul>
  </div>
  <div id="main">
    <h1><i>GenomeTools</i></h1>
    <p>
    The <i>GenomeTools</i> genome analysis system is a <a
    href="license.html">free</a> collection of bioinformatics tools (in the realm of
    genome informatics) combined into a single binary named <tt>gt</tt>.
    It is based on a C library named &ldquo;libgenometools&rdquo; which consists of
    several modules. For example, the <a
      href="annotationsketch.html"><tt>AnnotationSketch</tt></a> module.
    </p>
    <p>
    If you are interested in gene prediction, have a look at
    <a href="http://genomethreader.org" title="GenomeThreader gene prediction
      software"><i>GenomeThreader</i></a>.
    </p>
    <h2>Platforms</h2>
    <p>
    <i>GenomeTools</i> has been designed to run on every
    <a href="http://en.wikipedia.org/wiki/POSIX">POSIX</a> compliant
    <a href="http://en.wikipedia.org/wiki/UNIX">UNIX</a> system. For example, Linux,
    Mac OS X, and OpenBSD. An <strong>untested</strong>
    <a href="pub/windows">Windows version</a> (without
    <tt>AnnotationSketch</tt>) which runs under
    <a href="http://www.cygwin.com/">Cygwin</a> was kindly provided by
    Mohamed Abouelhoda. Please report problems with it to our
    <a href="mailman/listinfo/gt-users">mailing list</a>.
    </p>
    <p>
    <h2>Published software in <em>GenomeTools</em></h2>
    <p>
    The <em>GenomeTools</em> distribution includes several published software
    tools:
    </p>
    <ul class="pubs">
      <li><tt><a href="http://www.zbh.uni-hamburg.de/LTRharvest">ltrharvest</a></tt>, an efficient and flexible software tool for
      <em>de novo</em> detection of LTR retrotransposons.
      <div class="ref">
       D. Ellinghaus, S. Kurtz, and U. Willhoeft.<br>
       <i>LTRharvest</i>, a efficient and flexible software for <i> de novo </i>
       detection of LTR retrotransposons. <br>
       <a href="http://www.biomedcentral.com/1471-2105/9/18/abstract">
         BMC Bioinformatics 2008, 9:18
       </a>
      </div>
      </li>
      <li><tt><a href="http://www.zbh.uni-hamburg.de/tallymer">tallymer</a></tt>, a collection of flexible and memory-efficient
      programs for <em>k</em>-mer counting and indexing of large sequence sets.
      <div class="ref">
       S. Kurtz, A. Narechania, J.C. Stein, and D. Ware.<br>
       A new method to compute K-mer frequencies and its application to annotate
        large repetitive plant genomes.<br>
       <a href="http://www.biomedcentral.com/1471-2164/9/517/abstract">
         BMC Genomics 2008, 9:517
       </a>
      </div>
      </li>
      <li><tt>uniquesub</tt>, a program for computing minimum unique substrings.
      <div class="ref">
       S. Gr&auml;f, F.G.G. Nielsen, S. Kurtz, M.A. Huynen, E. Birney,
       H. Stunnenberg, and P. Flicek.<br>
       Optimized design and assessment of whole genome tiling arrays.<br>
       <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/23/13/i195">
         Bioinformatics 2007, 23(13):i195&ndash;i204
       </a>
      </div>
      </li>
      <li><tt>AnnotationSketch</tt>, a library for drawing genome annotations.
      <div class="ref">
       S. Steinbiss, G. Gremme, C. Sch&auml;rfer, M. Mader and S. Kurtz.<br>
       <em>AnnotationSketch</em>: a genome annotation drawing library.<br>
       <a href="http://bioinformatics.oxfordjournals.org/cgi/content/short/25/4/533">
         Bioinformatics 2009, 25(4):533&ndash;534
       </a>
      </div>
      </li>
    </ul>
    <h2>Visualization</h2>
    <p>
    With the <tt>gt sketch</tt> tool (build with
    <a href="annotationsketch.html"><tt>AnnotationSketch</tt></a>) you can produce PNG
    representations out of <a href="http://song.sourceforge.net/gff3.shtml"
      title="GFF3 format specification">GFF3</a>
    annotation files. The picture below was produced with the following command:
    </p>
    <pre>
    gt sketch -start 4000 -end 9000 annotation.png <a
    href="images/annotation.gff3" title="annotation file used to produce the picture">annotation.gff3</a>
    </pre>
    <p>
    <img src="images/annotation.png" alt="[Genome Annotation]">
    </p>
<div id="footer">
Copyright &copy; 2006-2009
<a href="mailto:gremme@gmail.com">Gordon Gremme.</a> Last update: 2009-01-28
</div>
  </div>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-301256-4";
urchinTracker();
</script>
</body>
</html>
